crates/ruff_python_parser/src/lexer.rs

```diff
@@ -1318,6 +1354,41 @@ impl<'source> Lexer<'source> {
     fn token_start(&self) -> TextSize {
         self.token_range().start()
     }
+
+    pub(crate) fn take_value(&mut self) -> TokenValue {
+        let (tok, _) = std::mem::replace(&mut self.current, (Tok::EndOfFile, TextRange::default()));
+        TokenValue::from_token(tok)
+    }
+
+    /// Creates a checkpoint to which it can later return to using [Self::rewind].
+    pub(crate) fn checkpoint(&self) -> LexerCheckpoint<'src> {
+        LexerCheckpoint {
```

Comment:
I think we also need to store the current token or `.current()` will return a stale token after rewinding.

Reply:
But what if we do speculative parsing? In that case the parser might call into `token_value` or `current` to get the kind of the current token.

Reply:
Yeah, in that case we'd need to store it. This would make caching lookahead a requirement